runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -40,76 +40,103 @@ const WASMER_FEATURES: wasmer::Features = wasmer::Features {
+        // `checked_sub` here verifies that offsetting the buffer by offset still lands us
+        // in-bounds of the allocated object.
+        let remaining = size.checked_sub(offset)?;
+        Some(unsafe {
+            // SAFETY: we verified that offsetting the base pointer by `offset` still lands us
+            // in-bounds of the original object.
+            (self.0.vmmemory().as_ref().base.add(offset), remaining)
+        })
+    }
+
+    fn get_memory_buffer(&self, offset: u64, len: usize) -> *mut u8 {
+        let memory = self.data_offset(offset).map(|(data, remaining)| (data, len <= remaining));
+        if let Some((ptr, true)) = memory {
+            ptr
+        } else {
+            panic!("memory access out of bounds")
+        }
+    }
+
+    pub(crate) fn vm(&self) -> VMMemory {
+        VMMemory { from: self.0.clone(), instance_ref: None }
     }
 }
 
 impl MemoryLike for Wasmer2Memory {
     fn fits_memory(&self, offset: u64, len: u64) -> bool {
-        match offset.checked_add(len) {
-            None => false,
-            Some(end) => self.0.size().bytes() >= Bytes(end as usize),
-        }
+        self.data_offset(offset)
+            .and_then(|(_, remaining)| {
+                let len = usize::try_from(len).ok()?;
+                Some(len <= remaining)
+            })
+            .unwrap_or(false)
     }
 
     fn read_memory(&self, offset: u64, buffer: &mut [u8]) {
-        let offset = offset as usize;
-        for (i, cell) in self.0.view()[offset..(offset + buffer.len())].iter().enumerate() {
-            buffer[i] = cell.get();
+        unsafe {
+            let memory = self.get_memory_buffer(offset, buffer.len());
+            // SAFETY: we verified indices into are valid and the pointer will always be valid as
+            // well. Our runtime is currently only executing Wasm code on a single thread, so data
+            // races aren't a concern here.
+            std::ptr::copy_nonoverlapping(memory, buffer.as_mut_ptr(), buffer.len());
         }
     }
 
     fn read_memory_u8(&self, offset: u64) -> u8 {
-        self.0.view()[offset as usize].get()
+        unsafe { *self.get_memory_buffer(offset, 1) }
     }
 
     fn write_memory(&mut self, offset: u64, buffer: &[u8]) {
-        let offset = offset as usize;
-        self.0.view()[offset..(offset + buffer.len())]
-            .iter()
-            .zip(buffer.iter())
-            .for_each(|(cell, v)| cell.set(*v));
+        unsafe {
+            let memory = self.get_memory_buffer(offset, buffer.len());
+            // SAFETY: we verified indices into are valid and the pointer will always be valid as
+            // well. Our runtime is currently only executing Wasm code on a single thread, so data
+            // races aren't a concern here.
+            std::ptr::copy_nonoverlapping(buffer.as_ptr(), memory, buffer.len());
+        }
     }
 }
 
-impl IntoVMError for wasmer::InstantiationError {
```

Comment:
No strong feelings, but I'd probably stick with `wasmer_engine::InstantiationError` here -- when translating errors, its useful to know at a glance which crate the error comes from.